tests/mockserver_tests/test_request_id_header.py

```diff
@@ -0,0 +1,317 @@
+                (1, REQ_RAND_PROCESS_ID, client_id, channel_id, 10, 1),
+            ),
+            (
+                "/google.spanner.v1.Spanner/ExecuteStreamingSql",
+                (1, REQ_RAND_PROCESS_ID, client_id, channel_id, 11, 1),
+            ),
+            (
+                "/google.spanner.v1.Spanner/ExecuteStreamingSql",
+                (1, REQ_RAND_PROCESS_ID, client_id, channel_id, 12, 1),
+            ),
+        ]
+        assert got_stream_segments == want_stream_segments
+
+    def test_database_run_in_transaction_retries_on_abort(self):
+        counters = dict(aborted=0)
+        want_failed_attempts = 2
+
+        def select_in_txn(txn):
+            results = txn.execute_sql("select 1")
+            for row in results:
+                _ = row
+
+            if counters["aborted"] < want_failed_attempts:
+                counters["aborted"] += 1
+                add_error(SpannerServicer.Commit.__name__, aborted_status())
+
+        add_select1_result()
+        if not getattr(self.database, "_interceptors", None):
+            self.database._interceptors = MockServerTestBase._interceptors
+
+        self.database.run_in_transaction(select_in_txn)
+
+    def test_database_execute_partitioned_dml_request_id(self):
+        add_select1_result()
+        if not getattr(self.database, "_interceptors", None):
+            self.database._interceptors = MockServerTestBase._interceptors
+        _ = self.database.execute_partitioned_dml("select 1")
+
+        requests = self.spanner_service.requests
+        self.assertEqual(3, len(requests), msg=requests)
+        self.assertTrue(isinstance(requests[0], BatchCreateSessionsRequest))
+        self.assertTrue(isinstance(requests[1], BeginTransactionRequest))
+        self.assertTrue(isinstance(requests[2], ExecuteSqlRequest))
+
+        # Now ensure monotonicity of the received request-id segments.
+        got_stream_segments, got_unary_segments = self.canonicalize_request_id_headers()
+        NTH_CLIENT = self.database._nth_client_id
+        CHANNEL_ID = self.database._channel_id
+        want_unary_segments = [
+            (
+                "/google.spanner.v1.Spanner/BatchCreateSessions",
+                (1, REQ_RAND_PROCESS_ID, NTH_CLIENT, CHANNEL_ID, 1, 1),
+            ),
+            (
+                "/google.spanner.v1.Spanner/BeginTransaction",
+                (1, REQ_RAND_PROCESS_ID, NTH_CLIENT, CHANNEL_ID, 2, 1),
+            ),
+        ]
+        want_stream_segments = [
+            (
+                "/google.spanner.v1.Spanner/ExecuteStreamingSql",
+                (1, REQ_RAND_PROCESS_ID, NTH_CLIENT, CHANNEL_ID, 3, 1),
+            )
+        ]
+
+        assert got_unary_segments == want_unary_segments
+        assert got_stream_segments == want_stream_segments
+
+    def test_unary_retryable_error(self):
+        add_select1_result()
+        add_error(SpannerServicer.BatchCreateSessions.__name__, unavailable_status())
```

Comment:
I changed the error code here from `INTERNAL` to `UNAVAILABLE`, because:
1. `INTERNAL` was not retried before for unary RPCs, and this change should not add retries for that. The specific `INTERNAL` errors that are retried are only relevant for streaming RPCs.
2. This PR seems to try to add additional retry handling for `UNAVAILABLE` errors. That handling is however never used, because the RPC itself is already retried by Gax, and this happens at a lower level. This again causes this test to seem to work, however the attempt number is not being increased during these retries. Because this test initially used `INTERNAL` to test, the additional retry handler was being invoked, and that did increase the attempt number.